### PR TITLE
[SPARK-39481][SQL] Do not push down complex filter condition

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

This PR enhances `PushPredicateThroughNonJoin` to make it do not push down complex filter condition if current project does not contain other complex expressions and `grandChild` only contains `LeafNode`, `Filter` and `Project`. For example:
```scala
spark.sql("CREATE TABLE t1(a int, b int, c int) using parquet")
spark.sql("SELECT * FROM (SELECT a, b + 1 AS b1, c + 2 AS c2 FROM t1) t WHERE a > 1 AND b1 > 10").explain(true)
```
Before this PR:
```
== Optimized Logical Plan ==
Project [a#1, (b#2 + 1) AS b1#0, c#3]
+- Filter ((isnotnull(a#1) AND isnotnull(b#2)) AND ((a#1 > 1) AND ((b#2 + 1) > 10)))
   +- Relation default.t1[a#1,b#2,c#3] parquet
```

After this PR:
```
== Optimized Logical Plan ==
Filter (isnotnull(b1#0) AND (b1#0 > 10))
+- Project [a#1, (b#2 + 1) AS b1#0, c#3]
   +- Filter (isnotnull(a#1) AND (a#1 > 1))
      +- Relation default.t1[a#1,b#2,c#3] parquet
```

### Why are the changes needed?

Reduce complex expression evaluation.

1. Case from [SPARK-39481](https://issues.apache.org/jira/browse/SPARK-39481?focusedCommentId=17556219&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17556219):

   Before this PR:
   ```
   == Optimized Logical Plan ==
   Project [int_col#0, string_col#1, pythonUDF0#10 AS int_col_copy#5]
   +- ArrowEvalPython [copy_int_col(int_col#0)#4], [pythonUDF0#10], 200
      +- Project [int_col#0, string_col#1]
         +- Filter (pythonUDF0#9 >= 3)
            +- ArrowEvalPython [copy_int_col(int_col#0)#4], [pythonUDF0#9], 200
               +- LogicalRDD [int_col#0, string_col#1], false
   ```

   After this PR:
   ```
   == Optimized Logical Plan ==
   Project [int_col#0, string_col#1, pythonUDF0#9 AS int_col_copy#5]
   +- Filter (isnotnull(pythonUDF0#9) AND (pythonUDF0#9 >= 3))
      +- ArrowEvalPython [copy_int_col(int_col#0)#4], [pythonUDF0#9], 200
         +- LogicalRDD [int_col#0, string_col#1], false
   ```

2. Case from [SPARK-36290](https://github.com/apache/spark/pull/33522/files/4aebe0a19fdf84516e30ab917eb149d39c4d4ac6#r743625946):

   Before this PR:
   ```
   == Optimized Logical Plan ==
   Project [a#225, b#226, c#236, d#237]
   +- Join Inner, (CAST(udf(cast(a as string)) AS INT)#250 = CAST(udf(cast(c as string)) AS INT)#251)
      :- Project [_1#220 AS a#225, _2#221 AS b#226, cast(pythonUDF0#253 as int) AS CAST(udf(cast(a as string)) AS INT)#250]
      :  +- BatchEvalPython [udf(cast(_1#220 as string))], [pythonUDF0#253]
      :     +- Project [_1#220, _2#221]
      :        +- Filter isnotnull(cast(pythonUDF0#252 as int))
      :           +- BatchEvalPython [udf(cast(_1#220 as string))], [pythonUDF0#252]
      :              +- LocalRelation [_1#220, _2#221]
      +- Project [_1#231 AS c#236, _2#232 AS d#237, cast(pythonUDF0#255 as int) AS CAST(udf(cast(c as string)) AS INT)#251]
         +- BatchEvalPython [udf(cast(_1#231 as string))], [pythonUDF0#255]
            +- Project [_1#231, _2#232]
               +- Filter isnotnull(cast(pythonUDF0#254 as int))
                  +- BatchEvalPython [udf(cast(_1#231 as string))], [pythonUDF0#254]
                     +- LocalRelation [_1#231, _2#232]
   ```
   
   After this PR:
   ```
   == Optimized Logical Plan ==
   Project [a#225, b#226, c#236, d#237]
   +- Join Inner, (CAST(udf(cast(a as string)) AS INT)#250 = CAST(udf(cast(c as string)) AS INT)#251)
      :- Project [_1#220 AS a#225, _2#221 AS b#226, cast(pythonUDF0#252 as int) AS CAST(udf(cast(a as string)) AS INT)#250]
      :  +- Filter isnotnull(cast(pythonUDF0#252 as int))
      :     +- BatchEvalPython [udf(cast(_1#220 as string))], [pythonUDF0#252]
      :        +- LocalRelation [_1#220, _2#221]
      +- Project [_1#231 AS c#236, _2#232 AS d#237, cast(pythonUDF0#253 as int) AS CAST(udf(cast(c as string)) AS INT)#251]
         +- Filter isnotnull(cast(pythonUDF0#253 as int))
            +- BatchEvalPython [udf(cast(_1#231 as string))], [pythonUDF0#253]
               +- LocalRelation [_1#231, _2#232]
   ```

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit test.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -1728,6 +1740,16 @@ object PushPredicateThroughNonJoin extends Rule[LogicalPlan] with PredicateHelpe
...
   }
+
+  /**
+   * Do not push complex predicate for these LogicalPlan to reduce complex expression evaluation.
+   */
+  private def doNotPushComplexPredicate(plan: LogicalPlan): Boolean = plan match {
+    case _: LeafNode => true
+    case _: Filter => true
+    case _: Project => true
+    case _ => false
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1751]: Should not include `Join`. For example:
```scala
spark.sql("create table t1(a int, b string) using parquet")
spark.sql("create table t2(x int, y string) using parquet")

spark.sql("select * from (select b + 1 as new_b, * from (select * from t1 right join t2 on a = x) t ) where new_b > 1").explain(true)
```

If push down:
```
== Optimized Logical Plan ==
Project [(cast(b#2 as double) + 1.0) AS new_b#0, a#1, b#2, x#3, y#4]
+- Join Inner, (a#1 = x#3)
   :- Filter ((isnotnull(b#2) AND ((cast(b#2 as double) + 1.0) > 1.0)) AND isnotnull(a#1))
   :  +- Relation default.t1[a#1,b#2] parquet
   +- Filter isnotnull(x#3)
      +- Relation default.t2[x#3,y#4] parquet
```

If do not push down:
```
== Optimized Logical Plan ==
Filter (isnotnull(new_b#0) AND (new_b#0 > 1.0))
+- Project [(cast(b#2 as double) + 1.0) AS new_b#0, a#1, b#2, x#3, y#4]
   +- Join RightOuter, (a#1 = x#3)
      :- Filter isnotnull(a#1)
      :  +- Relation default.t1[a#1,b#2] parquet
      +- Relation default.t2[x#3,y#4] parquet
```

File: sql/core/src/test/scala/org/apache/spark/sql/ExplainSuite.scala [hunk 2]
```diff
@@ -645,26 +645,26 @@ class ExplainSuiteAE extends ExplainSuiteHelper with EnableAdaptiveExecutionSuit
         """
           |(2) Filter [codegen id : 2]
           |Input [1]: [id#xL]
-          |Condition : ((id#xL > Subquery subquery#x, [id=#x]) AND isnotnull((id#xL % 10)))
+          |Condition : (id#xL > Subquery subquery#x, [id=#x])
```
Comment (wangyum) [inline sql/core/src/test/scala/org/apache/spark/sql/ExplainSuite.scala:648]: Before this PR:
```
== Optimized Logical Plan ==
Project [key#2L, value#3L]
+- Join Inner, (key#2L = id#6L)
   :- Project [(id#0L % 10) AS key#2L, id#0L AS value#3L]
   :  +- Filter ((id#0L > scalar-subquery#8 []) AND isnotnull((id#0L % 10)))
   :     :  +- Aggregate [max(id#11L) AS max(id)#10L]
   :     :     +- Range (0, 10, step=1, splits=Some(2))
   :     +- Range (0, 100, step=1, splits=Some(2))
   +- Range (0, 10, step=1, splits=Some(2))
```

After this PR:
```
== Optimized Logical Plan ==
Project [key#2L, value#3L]
+- Join Inner, (key#2L = id#6L)
   :- Filter isnotnull(key#2L)
   :  +- Project [(id#0L % 10) AS key#2L, id#0L AS value#3L]
   :     +- Filter (id#0L > scalar-subquery#8 [])
   :        :  +- Aggregate [max(id#11L) AS max(id)#10L]
   :        :     +- Range (0, 10, step=1, splits=Some(2))
   :        +- Range (0, 100, step=1, splits=Some(2))
   +- Range (0, 10, step=1, splits=Some(2))
```

**Comment by wangyum** [comment]
cc @cloud-fan @HyukjinKwon @sigmod

**Comment by icexelloss** [comment]
Any chance we can move forward with this patch? We are seeing this in our production environment and would like to back port the patch.

**Comment by HyukjinKwon** [comment]
cc @cloud-fan FYI

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 3]
```diff
@@ -1729,6 +1750,16 @@ object PushPredicateThroughNonJoin extends Rule[LogicalPlan] with PredicateHelpe
       case _ => false
     }
   }
+
+  /**
+   * Do not push complex predicate for these LogicalPlan to reduce complex expression evaluation.
+   */
+  private def doNotPushComplexPredicate(plan: LogicalPlan): Boolean = plan.forall {
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1757]: I think we should rename like shouldPushPredicate or isComplexPredicate

**Comment by icexelloss** [comment]
Thanks @HyukjinKwon !

**Comment by cloud-fan** [comment]
This reminds me of `CollapseProject`. We need to make decisions between "replicate expensive expressions" vs "have more Project operators". "have more Project operators" is not a big deal so `CollapseProject` is very conservative and only collapse project if only cheap expressions are replicated. However, filter pushdown is a more complicated case as a predicate may reduce the input size dramatically, and replicating it is still worth it.

@wangyum can you put more effort to justify your heuristic in this PR?

**Comment by cloud-fan** [comment]
Note: Python UDF also means a separated operator (we can't evaluate it with Project), so a simple solution would be always disallow replicate Python UDFs when pushing down filters or merging filters.

**Pull request closed without merging**